cores/esp8266/core_esp8266_main.cpp

```diff
@@ -400,79 +400,230 @@ extern "C" void __disableWiFiAtBootTime (void)
+
+        // The system_partition_table_regist will not allow partitions to
+        // overlap. EEPROM_start is a good choice for phy_data overlay. The SDK
+        // does not need to know about EEPROM_start. So we can omit it from the
+        // table. The real EEPROM access is after user_init() begins long after
+        // the PHY_DATA read. So it should be safe from conflicts.
+        phy_data = EEPROM_start - 0x40200000u;
+
+        // For SDKs 3.0 builds, "sdk3_begin_phy_data_spoof and
+        // user_rf_cal_sector_set" starts and stops the spoofing logic in
+        // `core_esp8266_phy.cpp`.
+        extern void sdk3_begin_phy_data_spoof();
+        sdk3_begin_phy_data_spoof();
+
+        ld_config_chip_size = phy_data + 4096 * 5;
+
+        // -DALLOW_SMALL_FLASH_SIZE=1
+        // Allows for small flash-size builds targeted for multiple devices,
+        // commonly IoT, of varying flash sizes.
+        #if !defined(FLASH_MAP_SUPPORT) & !defined(ALLOW_SMALL_FLASH_SIZE)
+        // Note, system_partition_table_regist will only catch when the build
+        // flash size value set by the Arduino IDE Tools menu is larger than
+        // the firmware image value detected and updated on the fly by esptool.
+        if (flashchip->chip_size != ld_config_chip_size) {
+            // Stop to avoid possible stored flash data corruption. This
+            // mismatch will not occur with flash size selection "Mapping
+            // defined by Hardware and Sketch".
+            chip_sz_str = PSTR("Flash size mismatch, check that the build setting matches the device.\n");
+            continue;
+        }
+        #elif defined(ALLOW_SMALL_FLASH_SIZE) && !defined(FLASH_MAP_SUPPORT)
+        // Note, while EEPROM is confined to a smaller flash size, we are still
+        // placing RF_CAL and SYSTEM_PARAMETER at the end of flash. To prevent
+        // this, esptool or its equal needs to not update the flash size in the
+        // .bin image.
+        #endif
+
+        #if FLASH_MAP_SUPPORT & defined(DEBUG_ESP_PORT)
+        // I don't think this will ever fail. Everything traces back to the results of spi_flash_get_id()
+        if (flash_size != flashchip->chip_size) {
+            chip_sz_str = PSTR("Flash size mismatch, check that the build setting matches the device.\n");
+            continue;
+        }
+        #endif
+
+        // All the examples I find, show the partition table in the global address space.
+        static partition_item_t at_partition_table[] =
+        {
+            { SYSTEM_PARTITION_PHY_DATA,         phy_data,         0x1000 }, // type 5
+            { SYSTEM_PARTITION_RF_CAL,           rf_cal,           0x1000 }, // type 4
+            { SYSTEM_PARTITION_SYSTEM_PARAMETER, system_parameter, 0x3000 }, // type 6
+        };
+        _at_partition_table = at_partition_table;
+        _at_partition_table_sz = sizeof(at_partition_table) / sizeof(at_partition_table[0]);
```

Comment:
```suggestion
        _at_partition_table_sz = std::size(at_partition_table);
```

Reply:
I assume this will happen at C++ compile time and we can count on not needing C++ runtime init.

Reply:
It is, number of elements is part of the type info.
`template <typename T, size_t Size> constexpr size_t size(const T (&)[Size]) { return Size; }`